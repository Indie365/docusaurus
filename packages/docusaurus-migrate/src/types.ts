/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

export type RawData = {
  header?: string;
  content?: string;
};

export type Data = {
  metadata: {[key: string]: string};
  rawContent: string;
};
export interface DocusaurusConfig {
  baseUrl: string;
  favicon: string;
  tagline?: string;
  title: string;
  url: string;
  organizationName?: string;
  projectName?: string;
  githubHost?: string;
  plugins: Array<[string, {[key: string]: any}]>;
  themes?: [];
  presets: [
    [
      string,
      {
        docs: {[key: string]: any};
        blog: {[key: string]: any};
        theme: {[key: string]: any};
      },
    ],
  ];
  themeConfig: {
    navbar: {
      title?: string;
      logo: {
        src?: string;
      };
      image?: string;
      links: Array<object>;
    };
    footer: {
      links: Array<{
        title: string;
        items: Array<{
          label: string;
          to: string;
        }>;
      }>;
      copyright?: string;
      logo: {
        src?: string;
      };
    };
    algolia?: object;
  };
  customFields: {
    [key: string]: unknown;
  };
  scripts?: (
    | string
    | {
        src: string;
        [key: string]: unknown;
      }
  )[];
  stylesheets?: (
    | string
    | {
        href: string;
        [key: string]: unknown;
      }
  )[];
}

export type Config = {
  title?: string;
  tagline?: string;
  url?: string;
  baseUrl?: string;
  defaultVersionShown?: string;
  organizationName?: string;
  projectName?: string;
  noIndex?: string;
  headerLinks?: Array<any>;
  headerIcon?: string;
  favicon?: string;
  colors?: any;
  copyright?: string;
  editUrl?: string;
  users?: Array<object>;
  disableHeaderTitle?: string;
  disableTitleTagline?: string;
  separateCss?: Array<object>;
  footerIcon?: string;
  translationRecruitingLink?: string;
  algolia?: object;
  gaTrackingId?: string;
  highlight?: object;
  markdownPlugins?: Array<() => void>;
  scripts?: Array<{src: string; [key: string]: any} | string>;
  stylesheets?: Array<{href: string; [key: string]: any} | string>;
  facebookAppId?: string;
  facebookComments?: true;
  facebookPixelId?: string;
  twitter?: string;
  twitterUsername?: string;
  twitterImage?: string;
  ogImage?: string;
  cleanUrl?: boolean;
  scrollToTop?: boolean;
  scrollToTopOptions?: object;
};
