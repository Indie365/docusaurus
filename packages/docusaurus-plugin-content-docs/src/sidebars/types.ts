/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import type {
  DocMetadataBase,
  VersionMetadata,
  NumberPrefixParser,
  SidebarOptions,
} from '../types';

export type SidebarItemBase = {
  className?: string;
  customProps?: Record<string, unknown>;
};

export type SidebarItemDoc = SidebarItemBase & {
  type: 'doc' | 'ref';
  label?: string;
  id: string;
};

export type SidebarItemLink = SidebarItemBase & {
  type: 'link';
  href: string;
  label: string;
};

export type SidebarItemAutogenerated = SidebarItemBase & {
  type: 'autogenerated';
  dirName: string;
};

// The user-input configuration, before normalization
export type SidebarItemCategoryConfig = SidebarItemBase & {
  type: 'category';
  label: string;
  items: SidebarItemConfig[];
  collapsed?: boolean;
  collapsible?: boolean;
};

export type SidebarItemConfig =
  | SidebarItemDoc
  | SidebarItemLink
  | SidebarItemAutogenerated
  | SidebarItemCategoryConfig
  | string
  | SidebarCategoriesShorthand;

export type SidebarCategoriesShorthand = {
  [sidebarCategory: string]: SidebarItemConfig[];
};

export type SidebarConfig = SidebarCategoriesShorthand | SidebarItemConfig[];

// Sidebar given by user that is not normalized yet. e.g: sidebars.json
export type SidebarsConfig = {
  [sidebarId: string]: SidebarConfig;
};

// Still has 'autogenerated', which is handled in processing
export type NormalizedSidebarItemCategory = SidebarItemBase & {
  type: 'category';
  label: string;
  items: NormalizedSidebarItem[];
  collapsed: boolean;
  collapsible: boolean;
};

export type NormalizedSidebarItem =
  | SidebarItemDoc
  | SidebarItemLink
  | NormalizedSidebarItemCategory
  | SidebarItemAutogenerated;

export type NormalizedSidebar = NormalizedSidebarItem[];
export type NormalizedSidebars = {
  [sidebarId: string]: NormalizedSidebar;
};

export type SidebarItemCategory = SidebarItemBase & {
  type: 'category';
  label: string;
  items: SidebarItem[];
  collapsed: boolean;
  collapsible: boolean;
};

export type SidebarItem =
  | SidebarItemDoc
  | SidebarItemLink
  | SidebarItemCategory;

export type Sidebar = SidebarItem[];
export type SidebarItemType = SidebarItem['type'];
export type Sidebars = {
  [sidebarId: string]: Sidebar;
};

// Reduce API surface for options.sidebarItemsGenerator
// The user-provided generator fn should receive only a subset of metadatas
// A change to any of these metadatas can be considered as a breaking change
export type SidebarItemsGeneratorDoc = Pick<
  DocMetadataBase,
  'id' | 'frontMatter' | 'source' | 'sourceDirName' | 'sidebarPosition'
>;
export type SidebarItemsGeneratorVersion = Pick<
  VersionMetadata,
  'versionName' | 'contentPath'
>;

export type SidebarItemsGeneratorArgs = {
  item: SidebarItemAutogenerated;
  version: SidebarItemsGeneratorVersion;
  docs: SidebarItemsGeneratorDoc[];
  numberPrefixParser: NumberPrefixParser;
  options: SidebarOptions;
};
export type SidebarItemsGenerator = (
  generatorArgs: SidebarItemsGeneratorArgs,
) => Promise<SidebarItem[]>;

// Also inject the default generator to conveniently wrap/enhance/sort the default sidebar gen logic
// see https://github.com/facebook/docusaurus/issues/4640#issuecomment-822292320
export type SidebarItemsGeneratorOptionArgs = {
  defaultSidebarItemsGenerator: SidebarItemsGenerator;
} & SidebarItemsGeneratorArgs;
export type SidebarItemsGeneratorOption = (
  generatorArgs: SidebarItemsGeneratorOptionArgs,
) => Promise<SidebarItem[]>;

export type OrderMetadata = {
  previous?: string;
  next?: string;
  sidebar?: string;
};
