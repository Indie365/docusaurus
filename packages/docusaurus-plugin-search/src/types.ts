/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

export interface Node {
  /**
   * The variant of a node.
   */
  type: string;

  /**
   * Information from the ecosystem.
   */
  data?: unknown;

  /**
   * Location of a node in a source document.
   * Must not be present if a node is generated.
   */
  position?: unknown;

  [key: string]: unknown;
}

export interface DataNode {
  type: string;
  anchor?: string;
  body: string;
  heading?: string;
}

export interface pluginOptions {
  include: Array<string>;
  exclude: Array<string>;
}
