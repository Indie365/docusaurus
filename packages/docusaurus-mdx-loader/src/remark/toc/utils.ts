/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {toValue} from '../utils';
import type {Node} from 'unist';
import type {
  MdxjsEsm,
  // @ts-expect-error: TODO see https://github.com/microsoft/TypeScript/issues/49721
} from 'mdast-util-mdx';
import type {TOCHeading, TOCItem, TOCItems, TOCSlice} from './types';
import type {Program, SpreadElement, ImportDeclaration} from 'estree';

export function getImportDeclarations(program: Program): ImportDeclaration[] {
  return program.body.filter(
    (item): item is ImportDeclaration => item.type === 'ImportDeclaration',
  );
}

export const isImport = (child: Node): child is MdxjsEsm => {
  if (child.type === 'mdxjsEsm') {
    return (child as MdxjsEsm).value.startsWith('import');
  }
  return false;
};

export const hasImports = (index: number): boolean => index > -1;

export const isExport = (child: Node): child is MdxjsEsm => {
  if (child.type === 'mdxjsEsm') {
    return (child as MdxjsEsm).value.startsWith('export');
  }
  return false;
};

export function isMarkdownImport(
  importDeclaration: ImportDeclaration,
): boolean {
  const importPath = importDeclaration.source.value;
  return typeof importPath === 'string' && /\.mdx?$/.test(importPath);
}

export function findDefaultImportName(
  importDeclaration: ImportDeclaration,
): string | undefined {
  return importDeclaration.specifiers.find(
    (o: Node) => o.type === 'ImportDefaultSpecifier',
  )?.local.name;
}

function createTOCSliceAST(tocSlice: TOCSlice): SpreadElement {
  return {
    type: 'SpreadElement',
    argument: {type: 'Identifier', name: tocSlice.name},
  };
}

async function createTOCHeadingAST({heading}: TOCHeading) {
  const {toString} = await import('mdast-util-to-string');
  const {valueToEstree} = await import('estree-util-value-to-estree');

  const tocItem: TOCItem = {
    value: toValue(heading, toString),
    id: heading.data!.id!,
    level: heading.depth,
  };

  return valueToEstree(tocItem);
}

async function createTOCItemAST(tocItem: TOCItems[number]) {
  switch (tocItem.type) {
    case 'slice':
      return createTOCSliceAST(tocItem);
    case 'heading':
      return createTOCHeadingAST(tocItem);
    default: {
      throw new Error(`unexpected toc item type`);
    }
  }
}

export async function createTOCExportNode(
  name: string,
  tocItems: TOCItems,
): Promise<MdxjsEsm> {
  const tocItemArrayAST = await Promise.all(tocItems.map(createTOCItemAST));

  return {
    type: 'mdxjsEsm',
    value: '', // See https://github.com/facebook/docusaurus/pull/9684#discussion_r1457595181
    data: {
      estree: {
        type: 'Program',
        body: [
          {
            type: 'ExportNamedDeclaration',
            declaration: {
              type: 'VariableDeclaration',
              declarations: [
                {
                  type: 'VariableDeclarator',
                  id: {
                    type: 'Identifier',
                    name,
                  },
                  init: {
                    type: 'ArrayExpression',
                    elements: tocItemArrayAST,
                  },
                },
              ],
              kind: 'const',
            },
            specifiers: [],
            source: null,
          },
        ],
        sourceType: 'module',
      },
    },
  };
}
