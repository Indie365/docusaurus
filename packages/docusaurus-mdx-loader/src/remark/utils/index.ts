/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import escapeHtml from 'escape-html';
import toString from 'mdast-util-to-string';
import type {Node} from '@docusaurus/mdx-loader';

export function toValue(node?: Node): string {
  if (node && node.type) {
    switch (node.type) {
      case 'text':
        return escapeHtml(node.value);
      case 'heading':
        return node.children!.map(toValue).join('');
      case 'inlineCode':
        return `<code>${escapeHtml(node.value)}</code>`;
      case 'emphasis':
        return `<em>${node.children!.map(toValue).join('')}</em>`;
      case 'strong':
        return `<strong>${node.children!.map(toValue).join('')}</strong>`;
      case 'delete':
        return `<del>${node.children!.map(toValue).join('')}</del>`;
      default:
    }
  }

  return toString(node!);
}
