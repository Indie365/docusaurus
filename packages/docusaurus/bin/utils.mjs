/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

/**
 *
 * @param {string} value
 * @returns {boolean|number}
 */
export function normalizePollValue(value) {
  if (value === undefined || value === '') {
    return false;
  }

  const parsedIntValue = Number.parseInt(value, 10);
  if (!Number.isNaN(parsedIntValue)) {
    return parsedIntValue;
  }

  return value === 'true';
}
