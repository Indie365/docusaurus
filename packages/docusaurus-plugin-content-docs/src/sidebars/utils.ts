/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import type {
  Sidebars,
  Sidebar,
  SidebarItem,
  SidebarItemCategory,
  SidebarItemLink,
  SidebarItemDoc,
  SidebarItemType,
} from './types';
import {mapValues, difference} from 'lodash';
import {getElementsAround, toMessageRelativeFilePath} from '@docusaurus/utils';
import type {DocMetadataBase} from '../types';

export function transformSidebarItems(
  sidebar: Sidebar,
  updateFn: (item: SidebarItem) => SidebarItem,
): Sidebar {
  function transformRecursive(item: SidebarItem): SidebarItem {
    if (item.type === 'category') {
      return updateFn({
        ...item,
        items: item.items.map(transformRecursive),
      });
    }
    return updateFn(item);
  }
  return sidebar.map(transformRecursive);
}

function collectSidebarItemsOfType<
  Type extends SidebarItemType,
  Item extends SidebarItem & {type: SidebarItemType},
>(type: Type, sidebar: Sidebar): Item[] {
  function collectRecursive(item: SidebarItem): Item[] {
    const currentItemsCollected: Item[] =
      item.type === type ? [item as Item] : [];

    const childItemsCollected: Item[] =
      item.type === 'category' ? item.items.flatMap(collectRecursive) : [];

    return [...currentItemsCollected, ...childItemsCollected];
  }

  return sidebar.flatMap(collectRecursive);
}

export function collectSidebarDocItems(sidebar: Sidebar): SidebarItemDoc[] {
  return collectSidebarItemsOfType('doc', sidebar);
}
export function collectSidebarCategories(
  sidebar: Sidebar,
): SidebarItemCategory[] {
  return collectSidebarItemsOfType('category', sidebar);
}
export function collectSidebarLinks(sidebar: Sidebar): SidebarItemLink[] {
  return collectSidebarItemsOfType('link', sidebar);
}

export function collectSidebarsDocIds(
  sidebars: Sidebars,
): Record<string, string[]> {
  return mapValues(sidebars, (sidebar) => {
    return collectSidebarDocItems(sidebar).map((docItem) => docItem.id);
  });
}

export function createSidebarsUtils(
  sidebars: Sidebars,
  docsById: Record<string, DocMetadataBase>,
): {
  getFirstDocIdOfFirstSidebar: () => string | undefined;
  getSidebarNameByDocId: (docId: string) => string | null | undefined;
  getDocNavigation: (docId: string) => {
    sidebarName: string | undefined;
    previousId: string | undefined;
    nextId: string | undefined;
  };
  checkSidebarsDocIds: (validDocIds: string[], sidebarFilePath: string) => void;
} {
  const sidebarNameToDocIds = collectSidebarsDocIds(sidebars);
  const docIdToSidebarName: Record<string, string> = {};
  Object.keys(sidebars).forEach((sidebarName) => {
    const sidebarDocItems = collectSidebarDocItems(sidebars[sidebarName]);
    sidebarDocItems.forEach((item) => {
      if (item.displayThisSidebar) {
        docIdToSidebarName[item.id] = sidebarName;
      }
    });
  });

  function getFirstDocIdOfFirstSidebar(): string | undefined {
    return Object.values(sidebarNameToDocIds)[0]?.[0];
  }

  function getSidebarNameByDocId(docId: string): string | null | undefined {
    const {
      frontMatter: {
        displayed_sidebar: displayedSidebar = docIdToSidebarName[docId],
      },
    } = docsById[docId];
    return displayedSidebar;
  }

  function getDocNavigation(docId: string): {
    sidebarName: string | undefined;
    previousId: string | undefined;
    nextId: string | undefined;
  } {
    const sidebarName = getSidebarNameByDocId(docId);
    if (!sidebarName) {
      return {
        sidebarName: undefined,
        previousId: undefined,
        nextId: undefined,
      };
    }
    if (!Object.keys(sidebars).includes(sidebarName)) {
      throw new Error(
        `Doc with ID ${docId} wants to display sidebar ${sidebarName} but a sidebar with this name doesn't exist`,
      );
    }
    const docIds = sidebarNameToDocIds[sidebarName];
    const currentIndex = docIds.indexOf(docId);
    if (currentIndex === -1) {
      return {
        sidebarName,
        previousId: undefined,
        nextId: undefined,
      };
    }
    const {previous, next} = getElementsAround(docIds, currentIndex);
    return {
      sidebarName,
      previousId: previous,
      nextId: next,
    };
  }

  function checkSidebarsDocIds(validDocIds: string[], sidebarFilePath: string) {
    const allSidebarDocIds = Object.values(sidebarNameToDocIds).flat();
    const invalidSidebarDocIds = difference(allSidebarDocIds, validDocIds);
    if (invalidSidebarDocIds.length > 0) {
      throw new Error(
        `Invalid sidebar file at "${toMessageRelativeFilePath(
          sidebarFilePath,
        )}".
These sidebar document ids do not exist:
- ${invalidSidebarDocIds.sort().join('\n- ')}

Available document ids are:
- ${validDocIds.sort().join('\n- ')}`,
      );
    }
  }

  return {
    getFirstDocIdOfFirstSidebar,
    getSidebarNameByDocId,
    getDocNavigation,
    checkSidebarsDocIds,
  };
}
