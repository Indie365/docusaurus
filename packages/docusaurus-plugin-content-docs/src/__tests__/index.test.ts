/**
 * Copyright (c) 2017-present, Facebook, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import path from 'path';
import {validate} from 'webpack';
import {isMatch} from 'picomatch';
import fs from 'fs-extra';
import pluginContentDocs from '../index';
import {loadContext} from '@docusaurus/core/src/server/index';
import {applyConfigureWebpack} from '@docusaurus/core/src/webpack/utils';
import {RouteConfig} from '@docusaurus/types';
import {posixPath} from '@docusaurus/utils';

const createFakeActions = (routeConfigs: RouteConfig[], contentDir) => {
  return {
    addRoute: (config: RouteConfig) => {
      routeConfigs.push(config);
    },
    createData: async (name, _content) => {
      return path.join(contentDir, name);
    },
  };
};

test('site with wrong sidebar file', async () => {
  const siteDir = path.join(__dirname, '__fixtures__', 'simple-site');
  const context = loadContext(siteDir);
  const sidebarPath = path.join(siteDir, 'wrong-sidebars.json');
  const plugin = pluginContentDocs(context, {
    sidebarPath,
  });
  return plugin
    .loadContent()
    .catch(e =>
      expect(e).toMatchInlineSnapshot(
        `[Error: Improper sidebars file, document with id 'goku' not found.]`,
      ),
    );
});

describe('empty/no docs website', () => {
  const siteDir = path.join(__dirname, '__fixtures__', 'empty-site');
  const context = loadContext(siteDir);

  test('no files in docs folder', async () => {
    await fs.ensureDir(path.join(siteDir, 'docs'));
    const plugin = pluginContentDocs(context, {});
    const content = await plugin.loadContent();
    const {docsMetadata, docsSidebars} = content;
    expect(docsMetadata).toMatchInlineSnapshot(`Object {}`);
    expect(docsSidebars).toMatchInlineSnapshot(`Object {}`);

    const routeConfigs = [];
    const pluginContentDir = path.join(context.generatedFilesDir, plugin.name);
    const actions = createFakeActions(routeConfigs, pluginContentDir);

    await plugin.contentLoaded({
      content,
      actions,
    });

    expect(routeConfigs).toEqual([]);
  });

  test('docs folder does not exist', async () => {
    const plugin = pluginContentDocs(context, {path: '/path/does/not/exist/'});
    const content = await plugin.loadContent();
    expect(content).toBeNull();
  });
});

describe('simple website', () => {
  const siteDir = path.join(__dirname, '__fixtures__', 'simple-site');
  const context = loadContext(siteDir);
  const sidebarPath = path.join(siteDir, 'sidebars.json');
  const pluginPath = 'docs';
  const plugin = pluginContentDocs(context, {
    path: pluginPath,
    sidebarPath,
  });
  const pluginContentDir = path.join(context.generatedFilesDir, plugin.name);

  test('getPathToWatch', () => {
    const pathToWatch = plugin.getPathsToWatch();
    const matchPattern = pathToWatch.map(filepath =>
      posixPath(path.relative(siteDir, filepath)),
    );
    expect(matchPattern).not.toEqual([]);
    expect(matchPattern).toMatchInlineSnapshot(`
      Array [
        "docs/**/*.{md,mdx}",
        "sidebars.json",
      ]
    `);
    expect(isMatch('docs/hello.md', matchPattern)).toEqual(true);
    expect(isMatch('docs/hello.mdx', matchPattern)).toEqual(true);
    expect(isMatch('docs/foo/bar.md', matchPattern)).toEqual(true);
    expect(isMatch('docs/hello.js', matchPattern)).toEqual(false);
    expect(isMatch('docs/super.mdl', matchPattern)).toEqual(false);
    expect(isMatch('docs/mdx', matchPattern)).toEqual(false);
    expect(isMatch('sidebars.json', matchPattern)).toEqual(true);
    expect(isMatch('versioned_docs/hello.md', matchPattern)).toEqual(false);
    expect(isMatch('hello.md', matchPattern)).toEqual(false);
    expect(isMatch('super/docs/hello.md', matchPattern)).toEqual(false);
  });

  test('configureWebpack', async () => {
    const config = applyConfigureWebpack(
      plugin.configureWebpack,
      {
        entry: './src/index.js',
        output: {
          filename: 'main.js',
          path: path.resolve(__dirname, 'dist'),
        },
      },
      false,
    );
    const errors = validate(config);
    expect(errors.length).toBe(0);
  });

  test('content', async () => {
    const content = await plugin.loadContent();
    const {docsMetadata, docsSidebars} = content;
    expect(docsMetadata.hello).toEqual({
      id: 'hello',
      permalink: '/docs/hello',
      previous: {
        title: 'baz',
        permalink: '/docs/foo/baz',
      },
      sidebar: 'docs',
      source: path.join('@site', pluginPath, 'hello.md'),
      title: 'Hello, World !',
      description: 'Hi, Endilie here :)',
    });

    expect(docsMetadata['foo/bar']).toEqual({
      id: 'foo/bar',
      next: {
        title: 'baz',
        permalink: '/docs/foo/baz',
      },
      permalink: '/docs/foo/bar',
      sidebar: 'docs',
      source: path.join('@site', pluginPath, 'foo', 'bar.md'),
      title: 'Bar',
      description: 'This is custom description',
    });

    expect(docsSidebars).toMatchSnapshot();

    const routeConfigs = [];
    const actions = createFakeActions(routeConfigs, pluginContentDir);

    await plugin.contentLoaded({
      content,
      actions,
    });

    expect(routeConfigs).not.toEqual([]);
    expect(routeConfigs).toMatchSnapshot();
  });
});

/* TODO for versioning */
describe('versioned website lol', () => {
  const versionedSiteDir = path.join(
    __dirname,
    '__fixtures__',
    'versioned-site',
  );
  const context = loadContext(versionedSiteDir);
  const sidebarPath = path.join(versionedSiteDir, 'sidebars.json');
  const plugin = pluginContentDocs(context, {
    sidebarPath,
  });
  const pluginContentDir = path.join(context.generatedFilesDir, plugin.name);

  test('getPathToWatch', () => {
    const pathToWatch = plugin.getPathsToWatch();
    const matchPattern = pathToWatch.map(filepath =>
      posixPath(path.relative(versionedSiteDir, filepath)),
    );
    expect(matchPattern).not.toEqual([]);
    expect(matchPattern).toMatchInlineSnapshot(`
      Array [
        "docs/**/*.{md,mdx}",
        "versioned_sidebars/version-1.0.1-sidebars.json",
        "versioned_sidebars/version-1.0.0-sidebars.json",
        "versioned_docs/version-1.0.1/**/*.{md,mdx}",
        "versioned_docs/version-1.0.0/**/*.{md,mdx}",
        "sidebars.json",
      ]
    `);
    expect(isMatch('docs/hello.md', matchPattern)).toEqual(true);
    expect(isMatch('docs/hello.mdx', matchPattern)).toEqual(true);
    expect(isMatch('docs/foo/bar.md', matchPattern)).toEqual(true);
    expect(isMatch('sidebars.json', matchPattern)).toEqual(true);
    expect(
      isMatch('versioned_docs/version-1.0.0/hello.md', matchPattern),
    ).toEqual(true);
    expect(
      isMatch('versioned_docs/version-1.0.0/foo/bar.md', matchPattern),
    ).toEqual(true);
    expect(
      isMatch('versioned_sidebars/version-1.0.0-sidebars.json', matchPattern),
    ).toEqual(true);

    // Non existing version
    expect(
      isMatch('versioned_docs/version-2.0.0/foo/bar.md', matchPattern),
    ).toEqual(false);
    expect(
      isMatch('versioned_docs/version-2.0.0/hello.md', matchPattern),
    ).toEqual(false);
    expect(
      isMatch('versioned_sidebars/version-2.0.0-sidebars.json', matchPattern),
    ).toEqual(false);

    expect(isMatch('docs/hello.js', matchPattern)).toEqual(false);
    expect(isMatch('docs/super.mdl', matchPattern)).toEqual(false);
    expect(isMatch('docs/mdx', matchPattern)).toEqual(false);
    expect(isMatch('hello.md', matchPattern)).toEqual(false);
    expect(isMatch('super/docs/hello.md', matchPattern)).toEqual(false);
  });

  test('content', async () => {
    // const content = await plugin.loadContent();
    // const {docsMetadata, docsSidebars} = content;
    // TODO: expect docsMetadata.blabla.toEqual()
    // expect(docsSidebars).toMatchSnapshot();
    // const routeConfigs = [];
    // const actions = createFakeActions(routeConfigs, pluginContentDir);
    // await plugin.contentLoaded({
    //   content,
    //   actions,
    // });
    // expect(routeConfigs).not.toEqual([]);
    // expect(routeConfigs).toMatchSnapshot();
  });
});

test('bad docs folder contains reserved version folder', async () => {
  const badSiteDir = path.join(__dirname, '__fixtures__', 'bad-site');
  const context = loadContext(badSiteDir);
  const plugin = pluginContentDocs(context, {});
  await expect(plugin.loadContent()).rejects.toThrowErrorMatchingInlineSnapshot(
    `"You cannot have a folder named \\"version-1.0.0\\""`,
  );
});
