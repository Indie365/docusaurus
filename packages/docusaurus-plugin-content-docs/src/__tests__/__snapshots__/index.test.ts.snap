// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sidebar site with wrong sidebar content 1`] = `
"Invalid sidebar file at \\"packages/docusaurus-plugin-content-docs/src/__tests__/__fixtures__/simple-site/wrong-sidebars.json\\".
These sidebar document ids do not exist:
- goku

Available document ids are:
- foo/bar
- foo/baz
- headingAsTitle
- hello
- ipsum
- lorem
- rootAbsoluteSlug
- rootRelativeSlug
- rootResolvedSlug
- rootTryToEscapeSlug
- slugs/absoluteSlug
- slugs/relativeSlug
- slugs/resolvedSlug
- slugs/tryToEscapeSlug"
`;

exports[`simple website content 1`] = `
Object {
  "docs": Array [
    Object {
      "collapsed": true,
      "collapsible": true,
      "items": Array [
        Object {
          "collapsed": true,
          "collapsible": true,
          "items": Array [
            Object {
              "id": "foo/bar",
              "type": "doc",
            },
            Object {
              "id": "foo/baz",
              "type": "doc",
            },
          ],
          "label": "foo",
          "type": "category",
        },
        Object {
          "href": "https://github.com",
          "label": "Github",
          "type": "link",
        },
        Object {
          "id": "hello",
          "type": "ref",
        },
      ],
      "label": "Test",
      "type": "category",
    },
    Object {
      "collapsed": true,
      "collapsible": true,
      "items": Array [
        Object {
          "id": "hello",
          "type": "doc",
        },
      ],
      "label": "Guides",
      "type": "category",
    },
  ],
}
`;

exports[`simple website content 2`] = `
Object {
  "pluginName": Object {
    "pluginId": Object {
      "path": "/docs",
      "versions": Array [
        Object {
          "docs": Array [
            Object {
              "id": "foo/bar",
              "path": "/docs/foo/bar",
              "sidebar": "docs",
            },
            Object {
              "id": "foo/baz",
              "path": "/docs/foo/bazSlug.html",
              "sidebar": "docs",
            },
            Object {
              "id": "headingAsTitle",
              "path": "/docs/headingAsTitle",
              "sidebar": undefined,
            },
            Object {
              "id": "hello",
              "path": "/docs/",
              "sidebar": "docs",
            },
            Object {
              "id": "ipsum",
              "path": "/docs/ipsum",
              "sidebar": undefined,
            },
            Object {
              "id": "lorem",
              "path": "/docs/lorem",
              "sidebar": undefined,
            },
            Object {
              "id": "rootAbsoluteSlug",
              "path": "/docs/rootAbsoluteSlug",
              "sidebar": undefined,
            },
            Object {
              "id": "rootRelativeSlug",
              "path": "/docs/rootRelativeSlug",
              "sidebar": undefined,
            },
            Object {
              "id": "rootResolvedSlug",
              "path": "/docs/hey/rootResolvedSlug",
              "sidebar": undefined,
            },
            Object {
              "id": "rootTryToEscapeSlug",
              "path": "/docs/rootTryToEscapeSlug",
              "sidebar": undefined,
            },
            Object {
              "id": "slugs/absoluteSlug",
              "path": "/docs/absoluteSlug",
              "sidebar": undefined,
            },
            Object {
              "id": "slugs/relativeSlug",
              "path": "/docs/slugs/relativeSlug",
              "sidebar": undefined,
            },
            Object {
              "id": "slugs/resolvedSlug",
              "path": "/docs/slugs/hey/resolvedSlug",
              "sidebar": undefined,
            },
            Object {
              "id": "slugs/tryToEscapeSlug",
              "path": "/docs/tryToEscapeSlug",
              "sidebar": undefined,
            },
          ],
          "isLast": true,
          "label": "Next",
          "mainDocId": "hello",
          "name": "current",
          "path": "/docs",
        },
      ],
    },
  },
}
`;

exports[`site with custom sidebar items generator sidebarItemsGenerator is called with appropriate data 1`] = `
Object {
  "defaultSidebarItemsGenerator": [Function],
  "docs": Array [
    Object {
      "frontMatter": Object {},
      "id": "API/Core APIs/Client API",
      "sidebarPosition": 0,
      "source": "@site/docs/3-API/01_Core APIs/0 --- Client API.md",
      "sourceDirName": "3-API/01_Core APIs",
    },
    Object {
      "frontMatter": Object {},
      "id": "API/Core APIs/Server API",
      "sidebarPosition": 1,
      "source": "@site/docs/3-API/01_Core APIs/1 --- Server API.md",
      "sourceDirName": "3-API/01_Core APIs",
    },
    Object {
      "frontMatter": Object {},
      "id": "API/Extension APIs/Plugin API",
      "sidebarPosition": 0,
      "source": "@site/docs/3-API/02_Extension APIs/0. Plugin API.md",
      "sourceDirName": "3-API/02_Extension APIs",
    },
    Object {
      "frontMatter": Object {},
      "id": "API/Extension APIs/Theme API",
      "sidebarPosition": 1,
      "source": "@site/docs/3-API/02_Extension APIs/1. Theme API.md",
      "sourceDirName": "3-API/02_Extension APIs",
    },
    Object {
      "frontMatter": Object {},
      "id": "API/api-end",
      "sidebarPosition": 3,
      "source": "@site/docs/3-API/03_api-end.md",
      "sourceDirName": "3-API",
    },
    Object {
      "frontMatter": Object {},
      "id": "API/api-overview",
      "sidebarPosition": 0,
      "source": "@site/docs/3-API/00_api-overview.md",
      "sourceDirName": "3-API",
    },
    Object {
      "frontMatter": Object {
        "id": "guide1",
        "sidebar_position": 1,
      },
      "id": "Guides/guide1",
      "sidebarPosition": 1,
      "source": "@site/docs/Guides/z-guide1.md",
      "sourceDirName": "Guides",
    },
    Object {
      "frontMatter": Object {
        "id": "guide2",
      },
      "id": "Guides/guide2",
      "sidebarPosition": 2,
      "source": "@site/docs/Guides/02-guide2.md",
      "sourceDirName": "Guides",
    },
    Object {
      "frontMatter": Object {
        "id": "guide2.5",
        "sidebar_position": 2.5,
      },
      "id": "Guides/guide2.5",
      "sidebarPosition": 2.5,
      "source": "@site/docs/Guides/0-guide2.5.md",
      "sourceDirName": "Guides",
    },
    Object {
      "frontMatter": Object {
        "id": "guide3",
        "sidebar_position": 3,
      },
      "id": "Guides/guide3",
      "sidebarPosition": 3,
      "source": "@site/docs/Guides/guide3.md",
      "sourceDirName": "Guides",
    },
    Object {
      "frontMatter": Object {
        "id": "guide4",
      },
      "id": "Guides/guide4",
      "sidebarPosition": undefined,
      "source": "@site/docs/Guides/a-guide4.md",
      "sourceDirName": "Guides",
    },
    Object {
      "frontMatter": Object {
        "id": "guide5",
      },
      "id": "Guides/guide5",
      "sidebarPosition": undefined,
      "source": "@site/docs/Guides/b-guide5.md",
      "sourceDirName": "Guides",
    },
    Object {
      "frontMatter": Object {},
      "id": "getting-started",
      "sidebarPosition": 0,
      "source": "@site/docs/0-getting-started.md",
      "sourceDirName": ".",
    },
    Object {
      "frontMatter": Object {},
      "id": "installation",
      "sidebarPosition": 1,
      "source": "@site/docs/1-installation.md",
      "sourceDirName": ".",
    },
  ],
  "item": Object {
    "dirName": ".",
    "type": "autogenerated",
  },
  "numberPrefixParser": [Function],
  "options": Object {
    "sidebarCollapsed": true,
    "sidebarCollapsible": true,
  },
  "version": Object {
    "contentPath": "docs",
    "versionName": "current",
  },
}
`;
